## Problem Statement
A telecommunications company provides subscription-based services including mobile phone connectivity, internet access, and digital entertainment services such as streaming television and movies. Revenue is primarily generated through recurring customer subscriptions, with customers billed monthly or annually depending on their contract type and payment arrangement. As with most subscription-based businesses, long-term profitability depends heavily on retaining existing customers rather than continuously acquiring new ones.

In this context, customer churn refers to customers who discontinue their relationship with the company. Within the provided dataset, churn is explicitly captured as a binary target variable (Churn), where a value of “Yes” indicates that a customer has left the service and “No” indicates that the customer remains active. This framing positions churn prediction as a supervised classification problem, where historical customer data is used to learn patterns associated with customer attrition.
Churn is a critical business concern because acquiring new customers is significantly more expensive than retaining existing ones. When a customer churns, the company not only loses future subscription revenue but may also incur additional marketing and promotional costs to replace that customer. Early identification of customers who are at risk of churning allows the business to intervene proactively through targeted retention strategies such as contract adjustments, service improvements, or personalized incentives.

The objective of this analysis is to develop a predictive model that estimates the likelihood of customer churn based on demographic attributes, service usage patterns, contract details, and billing information. By leveraging these features, the model aims to identify high-risk customers early enough for the business to take preventive action, thereby reducing churn rates and protecting long-term revenue.


In [ ]:
# import libraries
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.linear_model import LogisticRegression

In [ ]:
#load dataset

df = pd.read_csv("/content/WA_Fn-UseC_-Telco-Customer-Churn 2.csv")
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [ ]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [ ]:
# from lose inspection "Total charges" dtype is object and should be converted to float though i noticed 11 empty space. Checked against tenure and i realise there is atrend so i didnot want to force drop it so i consider imputation of zero


df['TotalCharges'] = df['TotalCharges'].replace(' ', 0).astype(float)


In [ ]:
df["Churn"].value_counts() / 7034 * 100

Churn
No     73.557009
Yes    26.570941
Name: count, dtype: float64

We know we have an imbalance dataset


In [ ]:
df.info

<bound method DataFrame.info of       customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0     7590-VHVEG  Female              0     Yes         No       1   
1     5575-GNVDE    Male              0      No         No      34   
2     3668-QPYBK    Male              0      No         No       2   
3     7795-CFOCW    Male              0      No         No      45   
4     9237-HQITU  Female              0      No         No       2   
...          ...     ...            ...     ...        ...     ...   
7038  6840-RESVB    Male              0     Yes        Yes      24   
7039  2234-XADUH  Female              0     Yes        Yes      72   
7040  4801-JZAZL  Female              0     Yes        Yes      11   
7041  8361-LTMKD    Male              1     Yes         No       4   
7042  3186-AJIEK    Male              0      No         No      66   

     PhoneService     MultipleLines InternetService OnlineSecurity  ...  \
0              No  No phone service             DSL             No  ...   
1             Yes                No             DSL            Yes  ...   
2             Yes                No             DSL            Yes  ...   
3              No  No phone service             DSL            Yes  ...   
4             Yes                No     Fiber optic             No  ...   
...           ...               ...             ...            ...  ...   
7038          Yes               Yes             DSL            Yes  ...   
7039          Yes               Yes     Fiber optic             No  ...   
7040           No  No phone service             DSL            Yes  ...   
7041          Yes               Yes     Fiber optic             No  ...   
7042          Yes                No     Fiber optic            Yes  ...   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No          No          No              No  Month-to-month   
1                 Yes          No          No              No        One year   
2                  No          No          No              No  Month-to-month   
3                 Yes         Yes          No              No        One year   
4                  No          No          No              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7038              Yes         Yes         Yes             Yes        One year   
7039              Yes          No         Yes             Yes        One year   
7040               No          No          No              No  Month-to-month   
7041               No          No          No              No  Month-to-month   
7042              Yes         Yes         Yes             Yes        Two year   

     PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0                 Yes           Electronic check          29.85         29.85   
1                  No               Mailed check          56.95       1889.50   
2                 Yes               Mailed check          53.85        108.15   
3                  No  Bank transfer (automatic)          42.30       1840.75   
4                 Yes           Electronic check          70.70        151.65   
...               ...                        ...            ...           ...   
7038              Yes               Mailed check          84.80       1990.50   
7039              Yes    Credit card (automatic)         103.20       7362.90   
7040              Yes           Electronic check          29.60        346.45   
7041              Yes               Mailed check          74.40        306.60   
7042              Yes  Bank transfer (automatic)         105.65       6844.50   

      Churn  
0        No  
1        No  
2       Yes  
3        No  
4       Yes  
...     ...  
7038     No  
7039     No  
7040     No  
7041    Yes  
7042     No  

[7043 rows x 21 columns]>

In [ ]:
df.drop(["customerID", "gender"], axis = 1, inplace = True)

In [ ]:
df.head()

SeniorCitizen Partner Dependents  tenure PhoneService     MultipleLines  \
0              0     Yes         No       1           No  No phone service   
1              0      No         No      34          Yes                No   
2              0      No         No       2          Yes                No   
3              0      No         No      45           No  No phone service   
4              0      No         No       2          Yes                No   

  InternetService OnlineSecurity OnlineBackup DeviceProtection TechSupport  \
0             DSL             No          Yes               No          No   
1             DSL            Yes           No              Yes          No   
2             DSL            Yes          Yes               No          No   
3             DSL            Yes           No              Yes         Yes   
4     Fiber optic             No           No               No          No   

  StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No              No  Month-to-month              Yes   
1          No              No        One year               No   
2          No              No  Month-to-month              Yes   
3          No              No        One year               No   
4          No              No  Month-to-month              Yes   

               PaymentMethod  MonthlyCharges  TotalCharges Churn  
0           Electronic check           29.85         29.85    No  
1               Mailed check           56.95       1889.50    No  
2               Mailed check           53.85        108.15   Yes  
3  Bank transfer (automatic)           42.30       1840.75    No  
4           Electronic check           70.70        151.65   Yes

Feature Engineering

In [ ]:
# encode binary columnns

binary_cols = [ "Partner", "Dependents", "PhoneService", "OnlineSecurity", "OnlineBackup", "DeviceProtection", "TechSupport", "StreamingTV", "StreamingMovies", 'PaperlessBilling']

for col in binary_cols:
  df[col] = df[col].map({"Yes": 1, "No" : 0, 'No internet service': 0 })




In [ ]:
df.head()

SeniorCitizen  Partner  Dependents  tenure  PhoneService     MultipleLines  \
0              0        1           0       1             0  No phone service   
1              0        0           0      34             1                No   
2              0        0           0       2             1                No   
3              0        0           0      45             0  No phone service   
4              0        0           0       2             1                No   

  InternetService  OnlineSecurity  OnlineBackup  DeviceProtection  \
0             DSL               0             1                 0   
1             DSL               1             0                 1   
2             DSL               1             1                 0   
3             DSL               1             0                 1   
4     Fiber optic               0             0                 0   

   TechSupport  StreamingTV  StreamingMovies        Contract  \
0            0            0                0  Month-to-month   
1            0            0                0        One year   
2            0            0                0  Month-to-month   
3            1            0                0        One year   
4            0            0                0  Month-to-month   

   PaperlessBilling              PaymentMethod  MonthlyCharges  TotalCharges  \
0                 1           Electronic check           29.85         29.85   
1                 0               Mailed check           56.95       1889.50   
2                 1               Mailed check           53.85        108.15   
3                 0  Bank transfer (automatic)           42.30       1840.75   
4                 1           Electronic check           70.70        151.65   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes

In [ ]:
# Encode categorical columns

cat_col = ["MultipleLines", "InternetService", "Contract", "PaymentMethod"]
df = pd.get_dummies(df, columns = cat_col, drop_first = True)

In [ ]:
df

SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0                 0        1           0       1             0   
1                 0        0           0      34             1   
2                 0        0           0       2             1   
3                 0        0           0      45             0   
4                 0        0           0       2             1   
...             ...      ...         ...     ...           ...   
7038              0        1           1      24             1   
7039              0        1           1      72             1   
7040              0        1           1      11             0   
7041              1        1           0       4             1   
7042              0        0           0      66             1   

      OnlineSecurity  OnlineBackup  DeviceProtection  TechSupport  \
0                  0             1                 0            0   
1                  1             0                 1            0   
2                  1             1                 0            0   
3                  1             0                 1            1   
4                  0             0                 0            0   
...              ...           ...               ...          ...   
7038               1             0                 1            1   
7039               0             1                 1            0   
7040               1             0                 0            0   
7041               0             0                 0            0   
7042               1             0                 1            1   

      StreamingTV  ...  Churn  MultipleLines_No phone service  \
0               0  ...     No                            True   
1               0  ...     No                           False   
2               0  ...    Yes                           False   
3               0  ...     No                            True   
4               0  ...    Yes                           False   
...           ...  ...    ...                             ...   
7038            1  ...     No                           False   
7039            1  ...     No                           False   
7040            0  ...     No                            True   
7041            0  ...    Yes                           False   
7042            1  ...     No                           False   

      MultipleLines_Yes  InternetService_Fiber optic InternetService_No  \
0                 False                        False              False   
1                 False                        False              False   
2                 False                        False              False   
3                 False                        False              False   
4                 False                         True              False   
...                 ...                          ...                ...   
7038               True                        False              False   
7039               True                         True              False   
7040              False                        False              False   
7041               True                         True              False   
7042              False                         True              False   

      Contract_One year  Contract_Two year  \
0                 False              False   
1                  True              False   
2                 False              False   
3                  True              False   
4                 False              False   
...                 ...                ...   
7038               True              False   
7039               True              False   
7040              False              False   
7041              False              False   
7042              False               True   

      PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
0                                     False                           

In [ ]:
# Signals customers who just joined — often higher risk of early churn

df["is_new_customer"] = df["tenure"] ==0
df["is_new_customer"] = df["is_new_customer"].astype(int)

In [ ]:
df['tenure_group'] = pd.cut(df['tenure'], bins=[0, 12, 24, 48, 60, 72],
                            labels=['0-12','13-24','25-48','49-60','61-72'])
# Optional: one-hot encode tenure_group
df = pd.get_dummies(df, columns=['tenure_group'], drop_first=True)


In [ ]:
services = ['PhoneService', 'MultipleLines_Yes', 'InternetService_Fiber optic',
            'InternetService_No', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
            'TechSupport', 'StreamingTV', 'StreamingMovies']

df['num_services'] = df[services].sum(axis=1)


In [ ]:
df.head()

SeniorCitizen  Partner  Dependents  tenure  PhoneService  OnlineSecurity  \
0              0        1           0       1             0               0   
1              0        0           0      34             1               1   
2              0        0           0       2             1               1   
3              0        0           0      45             0               1   
4              0        0           0       2             1               0   

   OnlineBackup  DeviceProtection  TechSupport  StreamingTV  ...  \
0             1                 0            0            0  ...   
1             0                 1            0            0  ...   
2             1                 0            0            0  ...   
3             0                 1            1            0  ...   
4             0                 0            0            0  ...   

   Contract_Two year  PaymentMethod_Credit card (automatic)  \
0              False                                  False   
1              False                                  False   
2              False                                  False   
3              False                                  False   
4              False                                  False   

   PaymentMethod_Electronic check  PaymentMethod_Mailed check is_new_customer  \
0                            True                       False               0   
1                           False                        True               0   
2                           False                        True               0   
3                           False                       False               0   
4                            True                       False               0   

   tenure_group_13-24  tenure_group_25-48  tenure_group_49-60  \
0               False               False               False   
1               False                True               False   
2               False               False               False   
3               False                True               False   
4               False               False               False   

   tenure_group_61-72  num_services  
0               False             1  
1               False             3  
2               False             3  
3               False             3  
4               False             2  

[5 rows x 30 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 30 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   SeniorCitizen                          7043 non-null   int64  
 1   Partner                                7043 non-null   int64  
 2   Dependents                             7043 non-null   int64  
 3   tenure                                 7043 non-null   int64  
 4   PhoneService                           7043 non-null   int64  
 5   OnlineSecurity                         7043 non-null   int64  
 6   OnlineBackup                           7043 non-null   int64  
 7   DeviceProtection                       7043 non-null   int64  
 8   TechSupport                            7043 non-null   int64  
 9   StreamingTV                            7043 non-null   int64  
 10  StreamingMovies                        7043 non-null   int64  
 11  Pape

In [ ]:
#Define feature and target

X = df.drop("Churn", axis = 1)
y = df["Churn"]

In [ ]:
y

0        No
1        No
2       Yes
3        No
4       Yes
       ... 
7038     No
7039     No
7040     No
7041    Yes
7042     No
Name: Churn, Length: 7043, dtype: object

In [ ]:
#Define feature and target

df["Churn"] = df["Churn"].map({"Yes": 1, "No":0})


In [ ]:


X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)


In [ ]:
from sklearn.preprocessing import StandardScaler

# Only scale numeric columns
num_cols = ['tenure', 'MonthlyCharges', 'TotalCharges', 'num_services']
scaler = StandardScaler()
X_train[num_cols] = scaler.fit_transform(X_train[num_cols])
X_test[num_cols] = scaler.transform(X_test[num_cols])


In [ ]:

model = LogisticRegression(class_weight='balanced', solver='liblinear', max_iter=1000)
model.fit(X_train, y_train)


LogisticRegression(class_weight='balanced', max_iter=1000, solver='liblinear')

In [ ]:


y_pred = model.predict(X_test)
y_prob = model.predict_proba(X_test)[:, 1]

print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))
print("ROC-AUC:", roc_auc_score(y_test, y_prob))


[[737 298]
 [ 77 297]]
              precision    recall  f1-score   support

          No       0.91      0.71      0.80      1035
         Yes       0.50      0.79      0.61       374

    accuracy                           0.73      1409
   macro avg       0.70      0.75      0.71      1409
weighted avg       0.80      0.73      0.75      1409

ROC-AUC: 0.8426903304141157


## Business Interpretation

The model identifies most churners (high recall = 79%), so it’s useful for retention campaigns.

Some false positives (precision = 50%) → not all predicted churners will actually churn. That’s acceptable if retention incentives are cheap.

Non-churners are sometimes predicted as churners (FP = 298) → might result in unnecessary marketing spend.

## Observation:

High recall for churners (Yes = 0.79) → model is good at catching churners, which is what we care about in business.

Precision for churners is lower (0.50) → half of predicted churners are false alarms → may cost some unnecessary retention campaigns.